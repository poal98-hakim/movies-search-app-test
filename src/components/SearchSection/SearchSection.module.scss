
.errorValidation{
    color: red;
    font-weight: 900;
}