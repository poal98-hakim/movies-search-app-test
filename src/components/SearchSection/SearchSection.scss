
.error-validation{
    color: red;
    font-weight: 900;
}