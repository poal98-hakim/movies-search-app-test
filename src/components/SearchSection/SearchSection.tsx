import React, { useContext, useState } from "react";
import SearchIcon from "../../images/search-icon-yellow.png";
import KeyIcon from "../../images/key-icon.png";
import SearchBar from "./SearchBar/SearchBar";
import "./SearchSection.scss";
import MainContext from "../../contexts/MainContext";
import Search from "./Search/Search";

interface IProps{
  onSearch: (search: string, apiKey: string) => void;
  totalResults?: number;
  error: string;
}
const SearchSection : React.VFC<IProps> = ({ onSearch, totalResults, error }) => {
  const {search, apiKey, setSearch, setApiKey} = useContext(MainContext);
  const [isValidated, setIsValidated] = useState<boolean>(true);

  const validate = () => {
    if(!isValidated) {
      return <span className="error-validation">Please fill in all fields!</span>
    }
    if(error){
      return <span className="error-validation">{error}</span>
    }
    return null;
  }

  return (
    <form className="form-search flex f-direction-column f-align-center" onSubmit={event => {
        event.preventDefault();
        if(!search || !apiKey){
            setIsValidated(false);
            return;
        } 
        setIsValidated(true);
        onSearch(search, apiKey);
    }}>
        <SearchBar 
          icon={{
            src: SearchIcon,
            alt: "icon-search"
          }} 
          placeholder="Search*" 
          type="text"
          value={search}
          onChange={setSearch}
        />
        <SearchBar 
          icon={{
            src: KeyIcon,
            alt: "icon-key"
          }} 
          placeholder="Api Key*" 
          type="text"
          value={apiKey}
          onChange={setApiKey}
        />
        <Search totalResults={totalResults || null} />
        {validate()}
      </form>
  );
}

export default SearchSection;
