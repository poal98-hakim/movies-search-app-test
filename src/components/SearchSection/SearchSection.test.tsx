import SearchSection from "./SearchSection";
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'

const PROPS = {
  onSearch: jest.fn(),
  totalResults: 5,
  error: "Invalid key"
}

describe('SearchSection', () => {
  let props = {...PROPS};

  beforeEach(() => { //before each test render the component and reset the value of props
    props = {...PROPS};
    render(<SearchSection {...props} />);
  });
  
  it('renders correctly', () => {
    const { baseElement } = render(<SearchSection {...props} />);
    expect(baseElement).toMatchSnapshot()
  })

  it('loads search bars', () => {
    expect(screen.getAllByTestId("search-bar")).toHaveLength(2);
  })

  it('shows correctly keyword search bar', () => {
    expect(screen.getByAltText("icon-search")).toBeInTheDocument();
    expect(screen.getAllByRole("img")[0]).toHaveAttribute("src", expect.stringContaining("search-icon-yellow.png"));
  })

  it('shows correctly api key search bar', () => {
    expect(screen.getByAltText("icon-key")).toBeInTheDocument();
    expect(screen.getAllByRole("img")[1]).toHaveAttribute("src", expect.stringContaining("key-icon.png"));
  })

  it('loads correct inputs', () => {
    const searchBarsInput = screen.getAllByRole("textbox");
    expect(searchBarsInput).toHaveLength(2);
    expect(searchBarsInput[0]).toHaveAttribute("placeholder", "Search*");
    expect(searchBarsInput[0]).toHaveAttribute("type", "text");
    expect(searchBarsInput[1]).toHaveAttribute("placeholder", "Api Key*");
    expect(searchBarsInput[1]).toHaveAttribute("type", "text");
  })

  it('correctly fires handler on typing inputs', async () => {
    const searchBarsInput = screen.getAllByRole("textbox");

    userEvent.type(searchBarsInput[0], "{enter}");
    expect(props.onSearch).not.toHaveBeenCalled(); //does not fire the call if you don't fill both the inputs
  })


})
