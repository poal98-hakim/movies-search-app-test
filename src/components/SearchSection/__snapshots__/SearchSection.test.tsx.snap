// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSection renders correctly 1`] = `
<body>
  <div>
    <form
      class="form-search flex f-direction-column f-align-center"
    >
      <div
        class="search-bar-wrapper flex f-align-center f-justify-center"
        data-testid="search-bar"
      >
        <img
          alt="icon-search"
          src="search-icon-yellow.png"
          width="25"
        />
        <input
          placeholder="Search*"
          type="text"
          value=""
        />
      </div>
      <div
        class="search-bar-wrapper flex f-align-center f-justify-center"
        data-testid="search-bar"
      >
        <img
          alt="icon-key"
          src="key-icon.png"
          width="25"
        />
        <input
          placeholder="Api Key*"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex f-align-center"
      >
        <input
          class="button"
          type="submit"
          value="Search"
        />
        <span
          class="num-results-movies"
        >
          5
           results
        </span>
      </div>
      <span
        class="error-validation"
      >
        Invalid key
      </span>
    </form>
  </div>
  <div>
    <form
      class="form-search flex f-direction-column f-align-center"
    >
      <div
        class="search-bar-wrapper flex f-align-center f-justify-center"
        data-testid="search-bar"
      >
        <img
          alt="icon-search"
          src="search-icon-yellow.png"
          width="25"
        />
        <input
          placeholder="Search*"
          type="text"
          value=""
        />
      </div>
      <div
        class="search-bar-wrapper flex f-align-center f-justify-center"
        data-testid="search-bar"
      >
        <img
          alt="icon-key"
          src="key-icon.png"
          width="25"
        />
        <input
          placeholder="Api Key*"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex f-align-center"
      >
        <input
          class="button"
          type="submit"
          value="Search"
        />
        <span
          class="num-results-movies"
        >
          5
           results
        </span>
      </div>
      <span
        class="error-validation"
      >
        Invalid key
      </span>
    </form>
  </div>
</body>
`;
